### PR TITLE
Fix double dispatch on linux

### PR DESCRIPTION
**Pull request by davidfowl**
- Fixing a regression introduced into linux sockets workloads when removing double dispatching on windows. We only want to dispatch by default if we're running on the old windows IO threadpool.

New version of https://github.com/dotnet/aspnetcore/pull/43849

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs [hunk 1]
```diff
@@ -48,6 +48,14 @@ public SocketConnectionContextFactory(SocketConnectionFactoryOptions options, IL
         var maxWriteBufferSize = _options.MaxWriteBufferSize ?? 0;
         var applicationScheduler = options.UnsafePreferInlineScheduling ? PipeScheduler.Inline : PipeScheduler.ThreadPool;
 
+        // Socket callbacks run on the threads polling for IO if we're using the old thread windows thread pool
+        var dispatchSocketCallbacks = OperatingSystem.IsWindows() &&
+                                      (Environment.GetEnvironmentVariable("DOTNET_ThreadPool_UsePortableThreadPoolForIO") == "0" ||
+                                      Environment.GetEnvironmentVariable("COMPlus_ThreadPool_UsePortableThreadPoolForIO") == "0");
```
Comment (halter73) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:54]: What if someone sets this to "false". Would the runtime interpret that differently than we are here? This really isn't an app context switch?

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs [hunk 2]
```diff
@@ -48,6 +48,14 @@ public SocketConnectionContextFactory(SocketConnectionFactoryOptions options, IL
         var maxWriteBufferSize = _options.MaxWriteBufferSize ?? 0;
         var applicationScheduler = options.UnsafePreferInlineScheduling ? PipeScheduler.Inline : PipeScheduler.ThreadPool;
 
+        // Socket callbacks run on the threads polling for IO if we're using the old thread windows thread pool
```
Comment (adityamandaleeka) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:51]:
```suggestion
        // Socket callbacks run on the threads polling for IO if we're using the old Windows thread pool
```

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:54]: The runtime only supports 1 and 0 for these AFAIK. 
cc @kouvel

Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:54]: https://github.com/dotnet/runtime/blob/8f18a0dff889449ab62086134903f89c7325b13f/src/coreclr/inc/clrconfigvalues.h#L533

@jkotas @vitek-karas Is there a better way to read these settings from managed code?

**Comment by JamesNK** [comment]
Is there a way to unit test this? It's not easily visible when wrong, and we've had a regression here once. What can we do to avoid that happening again?

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs (hunk 1 above)
Comment (jkotas) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:54]: This is about as good as you can do (in some cases, the alternative is to use private reflection).

I assume that this is temporary hack for .NET 7 only. Correct?

**Comment by davidfowl** [comment]
Look for performance regressions (like the ones we have now 😄).

**Comment by jkotas** [comment]
> Look for performance regressions (like the ones we have now 😄).

Just curious - do you measure performance with the obsolete native threadpool impl?

**Comment by davidfowl** [comment]
> Just curious - do you measure performance with the obsolete native threadpool impl?

Today we don't have anything that explicitly measures it no.

**Comment by jkotas** [comment]
Ok, so the env checks for `ThreadPool_UsePortableThreadPoolForIO` are best effort, without any validation. I do not have a problem with it - just pointing it out.

File: src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs (hunk 1 above)
Comment (davidfowl) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:54]: Correct!

**Review by JamesNK — APPROVED** [review]

**Comment by davidfowl** [comment]
cc @Pilchie

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Comment by davidfowl** [comment]
https://github.com/orgs/dotnet/teams/aspnet-build Merge please. 🙏

**Comment by dougbu** [comment]
@davidfowl please use `@aspnet-admins` to request servicing merges

**Comment by davidfowl** [comment]
Got it

**Pull request merged**